Migrate contributing and test writing docs from wiki to sphinx (#363)
ssec/sift: doc/source/installation.rst

Unless otherwise documented below, it is recommended that the software
bundle be used for any novice users as it requires the least amount of
prior execution of conda or command line tools. Conda-based installation
is best for testers, developers, and those wishing to customize versions
of dependencies used by SIFT.

System Requirements
-------------------

SIFT works on Windows, Mac, and Linux, but depends on many complex
libraries which may make installation difficult on some platforms or systems.
It also makes heavy use of GPU (video card) processing via the PyOpenGL and VisPy
libraries. In general it is best to have a local physical system to run SIFT as opposed
to a virtual machine or remote server. Virtual machines and remote servers
commonly have limited ways of displaying OpenGL visualizations to another
machine.

While the below system resources should work they should be considered the
lower end of machines for running SIFT. In general the better GPU and faster
memory (RAM) the better experience you'll have with SIFT.

* Windows 10+ / Mac OS X >11.0 / Linux >= Rocky Linux 8
* 8GB RAM
* Disk space (preferrably on a SSD/NVMe drive) with 20GB+ available

> * GPU with 2GB VRAM and OpenGL 3+ support

Reviewer: Sorry to reiterate on this. From my last demos at various conferences, one of the first question is always: how will it perform on a simple laptop, without a GPU? My answer was usually: it's ok, as long as you don't load too many layers/time steps/large-array-datasets, as this is my experience on our EUM laptops. I'm just worried that someone looking for the specs and reading about the heavy GPU usage, might decide to not even try if she/he doesn't have one. 

How about pointing out that GPUs are optional, but highly recommended?
```suggestion
* GPU with 2GB VRAM and OpenGL 3+ support (optional, but highly recommended)
```

Author: I disagree. Users don't need a monster of a machine and that's fine that we lowered these limits, but we're talking about some high resolution satellite data and you've changed the default image behavior to require loading the entire image into GPU memory. Regardless of what the results of testing were (tiling didn't release memory, etc), this at least logically means that more GPU memory is going to be used. Telling people they don't need a good GPU or that an Intel CPU-integrated GPU is good enough is really risky in my opinion.

Users purchase machines based on these specs. I don't care if someone says "let's see if I can get it to run on my system" for a system they already have and that it works fine on their low-end system. They were going to do that testing anyway. I don't want users to buy new machines, especially if buying multiple for a training course, and have none of those machines work well.

Overall we should do more testing of how things really perform on lower end hardware.

Reviewer: Ok, I trust your opinion then :)

Installation
============

SIFT is provided either as Conda packages or as a compressed all-in-one
bundle (tarball/zip) of all dependencies with a single start script.
Unless otherwise documented below, it is recommended that the software
bundle be used for any novice users as it requires the least amount of
prior execution of conda or command line tools. Conda-based installation
is best for testers, developers, and those wishing to customize versions
of dependencies used by SIFT.

System Requirements
-------------------

> SIFT works on Windows, Mac, and Linux, but depends on many complex libraries
> and relies heavily on GPU (video card) processing via the PyOpenGL and VisPy
> libraries which may make installation difficult on some platforms or systems.
> In general it is best to have a local physical system to run SIFT as opposed
> to a virtual machine or remote server. Virtual machines and remote servers
> commonly have limited ways of displaying OpenGL visualizations to another
> machine.

Author: I would say "relies heavily on GPU" is a bit too strong of a statement, no? It seems to me that it works ok also on simpler systems without it. I'm just afraid that this may scare off potential users that have limited resources. 
```suggestion
SIFT works on Windows, Mac, and Linux, but depends on many complex
libraries which may make installation difficult on some platforms or systems.
It also makes heavy use of GPU (video card) processing via the PyOpenGL and VisPy
libraries. In general it is best to have a local physical system to run SIFT as opposed
to a virtual machine or remote server. Virtual machines and remote servers
commonly have limited ways of displaying OpenGL visualizations to another
machine.
```

While the below system resources should work they should be considered the
lower end of machines for running SIFT. In general the better GPU and faster
memory (RAM) the better experience you'll have with SIFT.

* Windows 10+ / Mac OS X >11.0 / Linux >= Rocky Linux 8
* 16GB RAM
* SSD/NVMe drive with 128GB+ available
* GPU with 4GB VRAM and OpenGL 3+ support

> * Data files to be loaded can require up to 1TB or more

Author: I'm a little worried about the 2GB VRAM, but ok.

Ultimately operating system support is limited to those supported by
conda-forge as this is the primary package repository where dependencies
are pulled from.

.. _bundle-install:

Application and Bundle
----------------------

Windows Bundle
^^^^^^^^^^^^^^

SIFT is distributed as a compressed `.zip` archive with all dependencies
provided internally. The bundle (.zip) can be downloaded from our FTP server
`here <https://bin.ssec.wisc.edu/pub/sift/dist/>`_. Once downloaded and the zip
file extracted, the file `SIFT.bat` can be double clicked to start SIFT.

By default SIFT caches files in a "workspace" located
at the user's

> ``/Users/<User>/AppData/Local/CIMSS-SSEC/SIFT/Cache/workspace`` directory.

Reviewer: ```suggestion
``\Users\<User>\AppData\Local\SIFT\Cache\workspace`` directory.
```

Author: I'm not sure this would render properly. I think this was written with `/` is because Python understands it, but I get your point.

Reviewer: Yeah, I think the most frequent use case here is probably the user copy-pasting (part of) the path in the Explorer, so it's just nice if it works then

Configuration files for the application are stored in the user's
``/Users/<User>/AppData/Roaming/SIFT/settings`` directory.

Linux Bundle
^^^^^^^^^^^^

SIFT is available as an all-in-one tarball (``.tar.gz``) for Rocky Linux 8+
systems. It can be downloaded from our FTP server
`here <https://bin.ssec.wisc.edu/pub/sift/dist/>`_. Once downloaded the files
must be extracted:

.. code-block:: bash

>    tar -xzf SIFT_1.1.2.tar.gz
> 
> This will create a ``SIFT`` directory. SIFT can then be run by doing:
> 
> .. code-block:: bash
> 
>    SIFT/SIFT.sh

Reviewer: ```suggestion
   tar -xzf SIFT_X.Y.Z.tar.gz

This will create a ``SIFT_X.Y.Z`` directory. SIFT can then be run by doing:

.. code-block:: bash

   SIFT_X.Y.Z/SIFT.sh
```

separate from ones for developing it. In detail, you should not install
*uwsift* but only *uwsift-devel-deps* into an development environment, since
otherwise the installed SIFT software may interfere with the version from
the sources. And vice versa.

This said, let's assume that ``MY_ENV`` denotes the respective environment
and ``SIFT_CHANNEL`` a Conda channel, where the SIFT packages can be
found [#f1]_, the following common steps should be performed to prepare a clean
environment for the desired task::

  %> conda create --name MY_ENV --channel conda-forge --strict-channel-priority python=3.10
  %> conda activate MY_ENV
  (MY_ENV)%> conda config --env --add channels conda-forge
  (MY_ENV)%> conda config --env --add channels SIFT_CHANNEL
  (MY_ENV)%> conda config --set channel_priority strict

.. rubric:: Footnotes

.. [#f1] You need to ask for the URL or name of this ``SIFT_CHANNEL``. If you
	 build packages yourself, the local build directory can be used as
	 this channel, by default it is ``~/conda-channels/uwsift/`` (see
	 :ref:`conda-packaging`)

For Users
^^^^^^^^^

Install the package *uwsift* into an environment called e.g. ``work`` and
prepared as described above::

>   (work)%> conda install uwsift

Reviewer: Oh, I see now that there's already a nicer conda installation guidance here: https://github.com/ssec/sift/wiki/installation-guide#python-package-installation, so maybe we can just bring this over?

>   (work)%> conda install uwsift

Author: Yes, that was going to be my next task.
